File: server/templates/spring-boot/base/src/main/java/com/devstarter/app/service/AuthService.java
package com.devstarter.app.service;

import com.devstarter.app.model.User;
import com.devstarter.app.security.JwtTokenProvider;
import jakarta.validation.ValidationException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import static org.springframework.http.HttpStatus.CONFLICT;
import static org.springframework.http.HttpStatus.UNAUTHORIZED;

@Service
public class AuthService {

    private final UserService userService;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    public AuthService(UserService userService, PasswordEncoder passwordEncoder, JwtTokenProvider jwtTokenProvider) {
        this.userService = userService;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenProvider = jwtTokenProvider;
    }

    public AuthResult register(String email, String password) {
        validateCredentials(email, password);

        if (userService.existsByEmail(email)) {
            throw new ResponseStatusException(CONFLICT, "Email already registered");
        }

        String passwordHash = passwordEncoder.encode(password);
        User user = userService.createUser(email, passwordHash);

        String token = jwtTokenProvider.createToken(user.getEmail(), user.getRole());
        return new AuthResult(user.getId(), user.getEmail(), token);
    }

    public AuthResult login(String email, String password) {
        validateCredentials(email, password);

        User user;
        try {
            user = userService.getByEmail(email);
        } catch (Exception ex) {
            throw new ResponseStatusException(UNAUTHORIZED, "Invalid credentials");
        }

        if (!passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new ResponseStatusException(UNAUTHORIZED, "Invalid credentials");
        }

        String token = jwtTokenProvider.createToken(user.getEmail(), user.getRole());
        return new AuthResult(user.getId(), user.getEmail(), token);
    }

    private void validateCredentials(String email, String password) {
        if (email == null || email.isBlank()) {
            throw new ValidationException("Email is required");
        }
        if (password == null || password.length() < 8) {
            throw new ValidationException("Password must be at least 8 characters");
        }
    }

    public record AuthResult(Long userId, String email, String token) {
    }
}

File: server/templates/flask/base/app/services/auth_service.py
from __future__ import annotations

from flask_jwt_extended import create_access_token, get_jwt_identity

from app.models.user import User
from app.services.user_service import UserService
from app.utils.response import ApiError


class AuthService:
    def __init__(self):
        self.user_service = UserService()

    def register(self, email: str, password: str) -> tuple[User, str]:
        user = self.user_service.create_user(email=email, password=password)
        token = create_access_token(identity=str(user.id))
        return user, token

    def login(self, email: str, password: str) -> tuple[User, str]:
        if not email:
            raise ApiError("Email is required", status_code=400)
        if not password:
            raise ApiError("Password is required", status_code=400)

        user = self.user_service.get_by_email(email)
        if not user or not user.check_password(password):
            raise ApiError("Invalid email or password", status_code=401)

        token = create_access_token(identity=str(user.id))
        return user, token

    def get_current_user(self) -> User | None:
        identity = get_jwt_identity()
        if identity is None:
            return None
        return self.user_service.get_by_id(int(identity))
